### PR TITLE
Scaffolding: Create default templates

### LINKED ISSUE #25473: Wrong model-namespace using in generated context
Hello,

if I try to scaffold an empty database to control it later by ef-versioning the generated context contains an using for the models-namespace. If the database is empty this namespace is empty as well and the compilation throws an error.

Steps to reproduce:

1. create an empty database
2. use "dotnet ef database scaffold" in an project
3. try to compile the generated code.

Ive fixed this on my site but an general fix should prevent the first error-message which appears directly after executing the scaffold command.

Thanks,
Michael

### LINKED ISSUE #25546: RevEng: Dedupe code in scaffolding code generator
See GenerateManyToMany method.

See https://github.com/dotnet/efcore/pull/25544#discussion_r689923149

Main issue is that current implementation assumed that we won't have nesting in entity type fluent APIs (since we never detect owned entities or TPT) but the join entity is configured using nesting via `UsingEntity` API.

### LINKED ISSUE #25547: RevEng: Avoid redundant code in many to many

### LINKED ISSUE #27087: When scaffolding, namespace isn't added for generic type parameter within an array
For example, a PostgreSQL column of type `CIDR[]` (array of IP subnets) is scaffolded as:

```c#
public ValueTuple<IPAddress, int>[] Subnets { get; set; }
```

But the System.Net namespace isn't used. System.Net is used when the value tuple isn't in an array, and also for IPAddress[] without the tuple (PostgreSQL type `inet[]`).

Flagged by @ajruckman in https://github.com/npgsql/efcore.pg/issues/2193

### LINKED ISSUE #28187: Style: Error SA1514	Element documentation header should be preceded by blank line
.NET 6.0.300 

```
dotnet add package Microsoft.EntityFrameworkCore.Design

dotnet add package Microsoft.EntityFrameworkCore.SqlServer

dotnet ef dbcontext scaffold "Server=VYPC\MISASME2022;Database=MISASME2022Sample;Trusted_Connection=True;" Microsoft.EntityFrameworkCore.SqlServer -o Models
```

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	SA1514	Element documentation header should be preceded by blank line	Domain	C:\Users\donhu\source\repos\FSH\src\Core\Domain\Entities\Catalog\Account.cs	223	Active
```
image

image

Related https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1517.md

### PR DESCRIPTION
**Pull request by bricelam**
This refactors our existing scaffolding code into T4 templates that we can both precompile to use as our default code generator and also ship somehow (probably `dotnet new`) as a starting point for users to customize.

Note, we're still reviewing and refining the templates. But the bulk of the changes are ready to go. I'll update this PR as work progresses, but we could also just do the fine-tuning in a later PR if we're ready to merge this one.

Part of #4038, part of #14545, resolves #25546, resolves #25547, fixes #27087, part of #27588, fixes #28187

### CODE REVIEW

File: test/EFCore.Design.Tests/Scaffolding/Internal/ModelCodeGeneratorTestBase.cs [hunk 1]
```diff
@@ -19,21 +29,23 @@ public abstract class ModelCodeGeneratorTestBase
...
         AddScaffoldingServices(services);
 
-        var generator = services.BuildServiceProvider(validateScopes: true)
-            .GetServices<IModelCodeGenerator>()
-            .Last(g => g is CSharpModelGenerator);
+        var generators = services.BuildServiceProvider(validateScopes: true)
+            .GetServices<IModelCodeGenerator>();
+        var generator = Random.Shared.Next() % 12 != 0
+            ? generators.Last(g => g is CSharpModelGenerator)
+            : generators.Last(g => g is TextTemplatingModelGenerator);
```
Comment (bricelam) [inline test/EFCore.Design.Tests/Scaffolding/Internal/ModelCodeGeneratorTestBase.cs:43]: 😈

**Comment by ajcvickers** [comment]
@bricelam Is this ready for review? Looks like you just pushed changes.

**Comment by bricelam** [comment]
Yes, it can be reviewed. I expect a few more tweaks to the DbContext template, but this could be merged to unblock others from working on scaffolding.

File: src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs [hunk 2]
```diff
@@ -58,4 +61,743 @@ public static IEnumerable<string> GetReverseEngineeringErrors(this IReadOnlyMode
...
+                    conventionKey.DeclaringEntityType,
+                    conventionKey.DeclaringEntityType.GetProperties()));
+
+    /// <summary>
+    /// Gets value indicating whether this index can be entirely reperesented by a data annotation.
+    /// </summary>
+    /// <param name="index">The index.</param>
+    /// <param name="annotationCodeGenerator">The provider's annotation code generator.</param>
+    /// <returns><see langword="true"/> if this index can be reperesented by a data annotation.</returns>
+    public static bool HasDataAnnotation(this IIndex index, IAnnotationCodeGenerator annotationCodeGenerator)
```
Comment (ajcvickers) [inline src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:83]: Is `HasDataAnnotation` the right name for "can be entirely reperesented by a data annotation?"

Comment (ajcvickers) [reply src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:83]: Also, spelling, apparently. :-)

Comment (bricelam) [reply src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:83]: Happy to change it. To the user, these usually appear in predicates like `Options.UseDataAnnotations && x.HasDataAnnotation`

File: src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs [hunk 3]
```diff
@@ -58,4 +61,743 @@ public static IEnumerable<string> GetReverseEngineeringErrors(this IReadOnlyMode
...
+
+    /// <summary>
+    /// Gets the data annotations to configure a navigation property.
+    /// </summary>
+    /// <param name="navigation">The navigation property.</param>
+    /// <param name="annotationCodeGenerator">The provider's annotation code generator.</param>
+    /// <returns>The data annotations.</returns>
+    public static IEnumerable<AttributeCodeFragment> GetDataAnnotations(this INavigation navigation, IAnnotationCodeGenerator annotationCodeGenerator)
+    {
+        if (navigation.IsOnDependent
```
Comment (ajcvickers) [inline src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:270]: Do we ever do `Required` on navs? Or is that always on the FK?

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs [hunk 4]
```diff
@@ -1,1172 +1,911 @@
...
+﻿// ------------------------------------------------------------------------------
+// <auto-generated>
+//     This code was generated by a tool.
+//     Runtime Version: 17.0.0.0
+//  
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+// ------------------------------------------------------------------------------
+namespace Microsoft.EntityFrameworkCore.Scaffolding.Internal
```
Comment (ajcvickers) [inline src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:10]: So this is now just the processed template, correct?

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.tt [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+    var providerCode = services.GetRequiredService<IProviderConfigurationCodeGenerator>();
+    var annotationCodeGenerator = services.GetRequiredService<IAnnotationCodeGenerator>();
+    var code = services.GetRequiredService<ICSharpHelper>();
+
+    var usings = new List<string>
+    {
+        "System",
+        "System.Collections.Generic",
+        "Microsoft.EntityFrameworkCore",
+        "Microsoft.EntityFrameworkCore.Metadata"
```
Comment (ajcvickers) [inline src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.tt:31]: These are `usings` we always need, right? What is it that we need `Metadata` for?

File: src/EFCore.Design/Scaffolding/Internal/CallContext.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// ReSharper disable once CheckNamespace
+namespace System.Runtime.Remoting.Messaging;
+
+internal sealed class CallContext
+{
+    public static object? LogicalGetData(string name)
```
Comment (ajcvickers) [inline src/EFCore.Design/Scaffolding/Internal/CallContext.cs:9]: What is this?

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:270]: Yes, better to do both on navs

File: src/EFCore.Design/Scaffolding/Internal/CallContext.cs (hunk 6 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Scaffolding/Internal/CallContext.cs:9]: Are you sure that you want to know the answer?

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs (hunk 4 above)
Comment (bricelam) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.cs:10]: Correct. Generated whenever the .tt file is saved

File: src/EFCore.Design/Scaffolding/Internal/CallContext.cs (hunk 6 above)
Comment (bricelam) [reply src/EFCore.Design/Scaffolding/Internal/CallContext.cs:9]: The T4 Preprocessor generates (unused) code that references it. This is just here to make the compiler happy.

File: src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs (hunk 2 above)
Comment (bricelam) [reply src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:83]: `HandledByDataAnnotations` might be a good alternative.

File: src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs (hunk 3 above)
Comment (bricelam) [reply src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:270]: Today, we always generate foreign key properties, and just let the IsNullable metadata on properties drive the generation of [Required]. Is there something we should change here now, or is this just something to consider when if we ever allow users to exclude foreign key properties from being scaffolded?

File: src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.tt (hunk 5 above)
Comment (bricelam) [reply src/EFCore.Design/Scaffolding/Internal/CSharpDbContextGenerator.tt]: Not *always* needed, but we put them here so we can just assume they'll be in scope when we're generating configuration. I'll see if I can clean it up a bit.

**Comment by ErikEJ** [comment]
@bricelam  wonder how EF Core Power Tools can take advantage of all this, as it does not attempt to load the users project in any way. 🤥

**Comment by bricelam** [comment]
@ErikEJ It should be able to work just fine without loading the user's assembly. If you just set `ModelCodeGenerationOptions.ProjectDir` to the user's project, the TextTemplatingModelGenerator service will discover and use the CodeTemplates\EFCore\DbContext.t4 and EntityType.t4 files.

**Comment by ErikEJ** [comment]
@bricelam Wonderful!

**Comment by bricelam** [comment]
@ErikEJ I'm also working on a `dotnet new` experience for acquiring the default templates in [bricelam:new](https://github.com/bricelam/efcore/compare/tt...new), but just like in the EF6 Power Tools, it could be nice to have a gesture for this in the EF Core Power Tools too.

**Comment by ErikEJ** [comment]
@bricelam Yes, I plan to include the default templates and Allow the user to drop them in the project just like with Handlebars templates today. And potentially offer more template sets over time for special use cases like UI modelbinding.

File: src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore.Design/Extensions/ScaffoldingModelExtensions.cs:273]: Seems fine for now.

**Pull request merged**